control/control.go

```diff
@@ -688,85 +674,80 @@ func (p *pluginControl) verifyPlugin(lp *loadedPlugin) error {
 	newPlugins := []core.SubscribedPlugin{}
 	var serrs []serror.SnapError
 	for _, r := range requested {
-		// get expanded namespaces from requested metrics
-		newNss, err := p.matchQueryToNamespaces(r.Namespace())
+		// get all metric types available in metricCatalog which fulfill the requested namespace and version (if ver <=0 the latest version will be taken)
+		newMetrics, err := p.metricCatalog.GetMetrics(r.Namespace(), r.Version())
 		if err != nil {
 			log.WithFields(log.Fields{
 				"_block": "control",
 				"action": "expanding-requested-metrics",
 				"query":  r.Namespace(),
+				"err":    err,
 			}).Error("error matching requested namespace with metric catalog")
-			serrs = append(serrs, err)
+			serrs = append(serrs, serror.New(err))
 			continue
 		}
 
 		if controlLogger.Level >= log.DebugLevel {
-			for _, n := range newNss {
+			for _, m := range newMetrics {
 				controlLogger.WithFields(log.Fields{
-					"_block": "control",
-					"ns":     n.String(),
-				}).Debug("Expanded namespaces found")
+					"_block":  "control",
+					"ns":      m.Namespace().String(),
+					"version": m.Version(),
+				}).Debug("Expanded namespaces found in metric catalog")
 			}
 		}
 
-		if len(newNss) > 0 {
-			for _, ns := range newNss {
-				// Get metric types from metric catalog
-				m, err := p.metricCatalog.Get(ns, r.Version())
-				if err != nil {
-					log.WithFields(log.Fields{
-						"_block":    "control",
-						"action":    "expanding requested metrics",
-						"namespace": ns.String(),
-						"version":   r.Version(),
-					}).Error("error retreiving metric given a namespace and version.")
-					serrs = append(serrs, serror.New(fmt.Errorf("error retreiving metric %s:%d", ns.String(), r.Version())))
-					continue
-				}
-				// in case config tree doesn't have any configuration for current namespace
-				// it's needed to initialize config, otherwise it will stay nil and panic later on
-				config := configTree.Get(ns.Strings())
-				if config == nil {
-					config = cdata.NewNode()
-				}
-				newMetrics = append(newMetrics, &metric{
-					namespace: m.Namespace(),
-					version:   m.Version(),
-					config:    config,
-				})
-
-				config = configTree.Get([]string{""})
-				if config == nil {
-					config = cdata.NewNode()
-				}
-				plugin := subscribedPlugin{
-					name:     m.Plugin.Name(),
-					typeName: m.Plugin.TypeName(),
-					version:  m.Plugin.Version(),
-					config:   config,
-				}
+		for _, mt := range newMetrics {
+			// in case config tree doesn't have any configuration for current namespace
+			// it's needed to initialize config, otherwise it will stay nil and panic later on
+			cfg := configTree.Get(mt.Namespace().Strings())
+			if cfg == nil {
+				cfg = cdata.NewNode()
+			}
+			// set config to metric
+			mt.config = cfg
+
+			// loaded plugin which exposes the metric
+			lp := mt.Plugin
+			key := lp.Key()
+
+			// groups metricTypes by a plugin.Key()
+			pmt, _ := newMetricsGroupedByPlugin[key]
```

Comment:
Please, look on snap/master [control/control.go#L1092](https://github.com/intelsdi-x/snap/blob/master/control/control.go#L1092) - this is exactly the same code. So, it definitely works and do not panic, because where there is no item in map under the key (for example map is empty), the empty `pmt` will be returned (but not nil). 

I take this fragment because there is no need to do `groupMetricByPlugin()` every time at each collection. It's enough to do that when metrics and collector are taken for subscription_group. I limited making changes in the rest of existing code if these are not really necessary, so I kept that fragment in form as it is on current master. 

But I agree that this could be changed in the future together with adding `newMetricTypes()` which will return empty metricTypes or even change `map[string]metricTypes` to `map[string]*metricTypes` could be also taken into consideration.

@IRCody, what do you thing about keeping that as it is? Is it acceptable for you?